Empire/Empire.cpp

```diff
@@ -1468,6 +1470,43 @@ int Empire::StockpileID(ResourceType res) const {
     }
 }
 
+int Empire::SourceID() const {
+    std::shared_ptr<const UniverseObject> good_source = Source();
+    return good_source ? good_source->ID() : INVALID_OBJECT_ID;
+}
+
+std::shared_ptr<const UniverseObject> Empire::Source() const {
+    // Make it safe to use with a null this pointer.
+    // This generates a compiler warning, but it enables the natural line of code of
+    // "GetEmpire(id)->Source()" without requiring an intervening null check
```

Comment:
Just have a GetSourceForEmpire(int id) function that can return nullptr. Not having a check that GetEmpire(id) didn't return nullptr is weird / misleading.